app/services/katello/event_daemon.rb

```diff
@@ -1,60 +1,99 @@
+      def settings
+        SETTINGS[:katello][:event_daemon]
       end
 
-      if Rails.env.production? && $PROGRAM_NAME.match(/dynflow/)
-        Rails.logger.info("This appears to be the Dynflow process; not starting Katello Event Daemon")
-        return false
+      def pid
+        return unless pid_file && File.exist?(pid_file)
+
+        File.open(pid_file) { |f| f.read.to_i }
       end
 
-      if started?
-        Rails.logger.info("Katello Event Daemon already running; not attempting startup")
-        return false
+      def pid_file
+        Rails.root.join('tmp', 'pids', 'katello_event_daemon.pid')
       end
 
-      !::Foreman.in_rake? && !Rails.env.test? && SETTINGS[:katello][:event_daemon_enabled]
-    end
+      def lock_file
+        Rails.root.join('tmp', 'katello_event_daemon.lock')
+      end
 
-    def self.start
-      if runnable?
-        services.values.each(&:run)
+      def write_pid_file
+        return unless pid_file
 
-        at_exit do
-          stop
+        File.open(pid_file, 'w') { |f| f.puts Process.pid }
+      end
+
+      def stop
+        return unless pid == Process.pid
+        services.values.each(&:close)
+        File.unlink(pid_file) if pid_file && File.exist?(pid_file)
+      end
+
+      def start(worker: false)
+        return if !runnable? || settings[:multiprocess] && !worker
+
+        lockfile = File.open(lock_file, 'r')
+        begin
+          lockfile.flock(File::LOCK_EX)
+          return if started? # ensure it wasn't started while we waited for the lock
+
+          start_services
+          write_pid_file
+
+          at_exit do
+            stop
+          end
+
+          Rails.logger.info("Katello event daemon started process=#{Process.pid} multiprocess=#{settings[:multiprocess]}")
+        ensure
+          lockfile.flock(File::LOCK_UN)
         end
+      end
 
-        @pid = Process.pid
+      def started?
+        Process.kill(0, pid)
+        true
+      rescue Errno::ESRCH, TypeError # process no longer exists or we had no PID cached
+        false
       end
-    end
 
-    def self.stop
-      # in puma clustered mode, we should only stop the daemon
-      # from the parent process (where it actually is running), not forked workers
-      return unless @pid == Process.pid
+      def start_services
+        services.values.each(&:run)
+      end
 
-      services.values.each(&:close)
+      def runnable?
+        return false if started? || !settings[:enabled] || ::Foreman.in_rake? || Rails.env.test?
 
-      @pid = nil
-    end
+        if defined?(Rails::Console)
+          Rails.logger.debug("Not starting Katello Event Daemon in the Rails Console")
+          return false
+        end
 
-    def self.started?
-      @pid.present?
-    end
+        if $PROGRAM_NAME.match(/spring/)
+          Rails.logger.debug("Spring is running in the environment; not starting Katello Event Daemon")
+          return false
+        end
+
+        if $PROGRAM_NAME.match(/dynflow/)
```

Comment:
@adamruzicka hey, are you aware of any better way to see if this code is running in the same process as dynflow?

Reply:
In this context by "dynflow" you mean an executor?

Reply:
I mean anywhere that dynflow is loading the rails application code (initializers being triggered, for example). Maybe that's the executor, but maybe that's only one such place?

Reply:
Reworked the code so I don't need to check this any more :)

Reply:
I'm glad you did, I couldn't really come up with a reliable way which would work in all scenarios